plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -3359,6 +3364,61 @@ public boolean delegateMaterializedViewRefreshToConnector(ConnectorSession sessi
         return hiveMaterializedViewMetadata.refreshMaterializedView(session, name);
     }
 
+    @Override
+    public Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        requireNonNull(session, "session is null");
+        requireNonNull(tableName, "tableName is null");
+        if (isHiveSystemSchema(tableName.getSchemaName())) {
+            return Optional.empty();
+        }
+        // we need to chop off any "$partitions" and similar suffixes from table name while querying the metastore for the Table object
+        TableNameSplitResult tableNameSplit = splitTableName(tableName.getTableName());
+        Optional<Table> table = metastore.getTable(new HiveIdentity(session), tableName.getSchemaName(), tableNameSplit.getBaseTableName());
+        if (table.isEmpty() || VIRTUAL_VIEW.name().equals(table.get().getTableType())) {
+            return Optional.empty();
+        }
+
+        Optional<CatalogSchemaTableName> catalogSchemaTableName = tableRedirectionsProvider.redirectTable(session, table.get());
+
+        // stitch back the suffix we cut off.
+        return catalogSchemaTableName.map(name -> new CatalogSchemaTableName(
+                name.getCatalogName(),
+                new SchemaTableName(
+                        name.getSchemaTableName().getSchemaName(),
+                        name.getSchemaTableName().getTableName() + tableNameSplit.getSuffix().orElse(""))));
+    }
+
+    private static TableNameSplitResult splitTableName(String tableName)
+    {
+        int metadataMarkerIndex = tableName.lastIndexOf('$');
+        return metadataMarkerIndex <= 0 ? // marker not found or at the beginning of tableName
+                new TableNameSplitResult(tableName, Optional.empty()) :
+                new TableNameSplitResult(tableName.substring(0, metadataMarkerIndex), Optional.of(tableName.substring(metadataMarkerIndex)));
+    }
+
+    private static class TableNameSplitResult
+    {
+        private final String baseTableName;
+        private final Optional<String> suffix;
```

Comment:
you can make it `String`, since `Optional.of("")` is equivalent to `Optional.empty()`